### PR TITLE
Mhc optimization: Rewrites einsums for better xla fusion

### PR DESCRIPTION
**Pull request by h1t35h**
# Description

XLA optimizations does not seem to gel well with einsum for cases where dimensions are being expanded. Leading to a ~10- 15% performance drop in the current implementation. This change implements expanded einsums with jnp.expand_dims leading to better XLA fusion.

Before (110us): 
image
After (97us):
image

# Tests

All existing tests passing.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/mhc.py [hunk 1]
```diff
@@ -269,7 +267,8 @@ def __call__(
         1.0,
         eps=1e-6,
     )
-    layer_input = jnp.einsum("bskd,bsk -> bsd", x, pre_mapping, precision=self.matmul_precision)
+    # bskd, bsk -> bsd
+    layer_input = jnp.sum(x * jnp.expand_dims(pre_mapping, axis=3), axis=2)
```
Comment (NuojCheng) [inline src/maxtext/layers/mhc.py:278]: for my understanding, do you see perf gain by this change? I thought `einsum` would be more performant.

Comment (h1t35h) [reply src/maxtext/layers/mhc.py:278]: Yes, even I found it a bit counter-intuitive einsum being more performant is what my understanding was as well. I think while being generally true for some cases like this one it seems better optimization is coming by moving away from it. Specifically, I can see the all-reduce overlap. I tested the changes individually by just changing the `einsums` and `for_iloop` but that does not seem to work I'm only seeing the All-Reduce overlap when both the changes are present. 

Just `for_iloop` replaced : 
image
Just `for_iloop` and `einsum` replaced: 
image

I don't have a very solid reasoning behind why of it.

Comment (dandragona) [reply src/maxtext/layers/mhc.py:278]: This is also somewhat surprising to me. I am curious if you can share the xprof? Also is this an isolation test or was this taken from your training job?

I am curious how this change generalizes to other shapes, i.e., is the speedup an artifact of your workload?

Comment (dandragona) [reply src/maxtext/layers/mhc.py:278]: Another option is you can try to use mhclite. It removes the need for the sinkhorn algorithm so theres no while loop. It also might be easier to fuse because it replace the sinkhorn step with an einsum.

Comment (dandragona) [reply src/maxtext/layers/mhc.py:278]: Upon further digging, I believe the rewrite to expand_dims being faster may be because dot_general is potentially harder to fuse than element wise operations. The contracting dimensions for H_pre, H_post are quite small (H_post is literally 1) so potentially doing elementwise operations is the move here.

I still think its worth a shot to see if mhclite helps.

Comment (dandragona) [reply src/maxtext/layers/mhc.py:278]: Here and elsewhere let's leave a comment about this theory and possible link to this discussion.

**Review by dandragona — APPROVED** [review]

File: src/maxtext/layers/mhc.py (hunk 1 above)
Comment (h1t35h) [reply src/maxtext/layers/mhc.py:278]: Done.

**Review by NuojCheng — APPROVED** [review]
